### PR TITLE
feat(diagnostic): add `suffix` option to `virt_text` config

### PR DESCRIPTION
**Pull request by beardedsakimonkey**
Follow-up to https://github.com/neovim/neovim/pull/21130

This introduces a `suffix` option to the `virt_text` config in `vim.diagnostic.config()`. The suffix can either be a string which is appended to the diagnostic message or a function returning such. The function receives a `diagnostic` argument, which is the diagnostic table of the last diagnostic (the one whose message is rendered as virt text).

We could also pass in `i` and `total` to the function, like we do for the `suffix` in `vim.diagnostic.open_float()`. This would enable configuring a suffix like " [3/3]" for a line with 3 diagnostics. I'm not sure how useful that is, though.

Also unlike the `suffix` in `vim.diagnostic.open_float()`, we don't support customizing the highlight. I'm not sure how useful this would be.

Example configuration:
```lua
local function suffix(diagnostic)
  return diagnostic.code and (" [%s]"):format(diagnostic.code) or "" 
end
vim.diagnostic.config({virtual_text = {suffix = suffix}})
```

Screenshot:
Screen Shot 2022-11-20 at 12 12 56 PM

cc @gpanders

### CODE REVIEW

**Comment by gpanders** [comment]
>We could also pass in i and total to the function, like we do for the suffix in vim.diagnostic.open_float(). This would enable configuring a suffix like " [3/3]" for a line with 3 diagnostics. I'm not sure how useful that is, though.

>Also unlike the suffix in vim.diagnostic.open_float(), we don't support customizing the highlight. I'm not sure how useful this would be.

For things like this, my policy is "just wait until someone asks for it" :)

File: runtime/doc/diagnostic.txt
Comment (gpanders) [inline runtime/doc/diagnostic.txt:370]: Add a link to `|diagnostic-structure|` somewhere in here so it's clear what the function argument is.

**Review by gpanders — APPROVED** [review]
Only one comment on documentation, otherwise LGTM.

Thank you for these pair of PRs! Good work.

**Comment by gpanders** [comment]
> We could also pass in `i` and `total` to the function, like we do for the `suffix` in `vim.diagnostic.open_float()`. This would enable configuring a suffix like " [3/3]" for a line with 3 diagnostics. I'm not sure how useful that is, though.
> 
> Also unlike the `suffix` in `vim.diagnostic.open_float()`, we don't support customizing the highlight. I'm not sure how useful this would be.

Can you remove these bits from the commit message? This is helpful info for the PR discussion, but doesn’t need to be captured in the commit itself.

File: runtime/lua/vim/diagnostic.lua
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:1064]: `prefix` does not require a trailing space, so `suffix` should not require a leading space, for consistency (unless there is a compelling argument or use case otherwise?).

Comment (beardedsakimonkey) [reply runtime/lua/vim/diagnostic.lua:1064]: Good point.  I guess one compelling reason is for consistency with `open_float()`'s suffix, which enables reusing the same `suffix` function for both virt_text and float.  But I don't feel strongly either way.

Comment (gpanders) [reply runtime/lua/vim/diagnostic.lua:1064]: That’s a valid reason, so let’s leave it as is. Ideally, `prefix` should not assume a trailing space either, but that ship has sailed.

**Comment by gpanders** [comment]
Ah I forgot to mention this before I merged, but we need entries in news.txt! Can you make one more PR mentioning the new suffix option for both virtual text and the floating window in `news.txt`?

File: runtime/lua/vim/diagnostic.lua
Comment (beardedsakimonkey) [reply runtime/lua/vim/diagnostic.lua:1064]: I think the reason `prefix` doesn't need a trailing space is because when a line has multiple diagnostics, it renders multiple prefixes (eg. `■■`)

**Comment by petobens** [comment]
Hi @beardedsakimonkey (and @gpanders)! Are there any plans to add `suffix` to `vim.diagnostic.setloclist` and ` vim.diagnostic.setqflist`? I would like to add `diagnostic.code` to each entry in such lists. Thanks in advance!!

**Comment by beardedsakimonkey** [comment]
Hi @petobens.  I didn't have any such plans, but only because I seldom use qf/location lists.

From what I can tell, it seems like there isn't really an appropriate field in the qflist data structure to put the error code into (see `:help setqflist-what`).  So, it seems like the suffix would need to be appended to the diagnostic message.

It seems like a reasonable feature to support, but, as someone who isn't very familiar with these features, I'd defer to the neovim team to decide if it's desirable.

**Pull request merged**
